File: test/test2.rb
m={}

class Obj
  attr_reader :hash

  ACONSTANT = 0

  def initialize(n)
    @val = n
    @hash = n
  end

  def to_s
    "val=#{@val}"
  end

  def func2
    num_tournaments.times do |i| # TODO: Find a way to appreciate the progress
      reproduction
      control
    end
  end
end

o1 = Obj.new(1)
o2 = Obj.new(1)
m[o1] = "hello"
m[o2] = "world"

p o1,o2
p m

# auto-add of parenthesis
p 9.modulo(3+2)
3.modulo(2).to_s()

# use of is_a?
p 3.2.is_a?(Float)
p 3.is_a?(Fixnum)
p "t".is_a?(String)
p [].is_a?(Array)

# use of gsub
p "abcbd".gsub("b", "x")
p "abcbd".gsub(/b/, "x")
p "abcbd".gsub(/B/i, "x")
p "abcbd".gsub(func2(), "x")

# call a block
def fnBlock(p1, &block)
  block.call
  block.call(3)
end

# slice or []
s = "abcdef"
p s[2]
p s[2,1]
p s.slice(2)
p s.slice(2,1)

def testLoops
  x = 2
  0.step(10,2) { |i| p i }
  0.step(-10,-2) { |i| p i }
  0.step(10,x) { |i3| p i3 }
  0.step(10,x) { |i3| p i3 }
  3.times { |j| p j }
  3.times { |j| p j }
  3.times { p x }
  x.each_key { |k| p x.keys }
end

File: test/js-out/Test2.js
//Translated from test2.rb using babyruby2js
'use strict';

var main = require('./main');
var m = {};

/** @class public read-only attribute: hash
 */
function Obj(n) {
    this.val = n;
    this.hash = n;
}
module.exports = Obj;

Obj.ACONSTANT = 0;
Obj.prototype.toString = function () {
    return 'val=' + this.val;
};

Obj.prototype.func2 = function () {
    for (var i = 0; i < numTournaments(); i++) { // TODO: Find a way to appreciate the progress
        reproduction();
        control();
    }
};

var o1 = new Obj(1);
var o2 = new Obj(1);
m[o1] = 'hello';
m[o2] = 'world';
console.log(o1, o2);
console.log(m);
// auto-add of parenthesis
console.log(9 % (3 + 2));
(3 % 2).toString();
// use of is_a?
console.log(main.isA('Float', 3.2));
console.log(main.isA('Fixnum', 3));
console.log(main.isA(String, 't'));
console.log(main.isA(Array, []));
// use of gsub
console.log('abcbd'.replaceAll('b', 'x'));
console.log('abcbd'.replace(/b/g, 'x'));
console.log('abcbd'.replace(/B/ig, 'x'));
console.log('abcbd'.replaceAll(func2(), 'x'));
// call a block
main.prototype.fnBlock = function (p1, block) {
    block();
    return block(3);
};

// slice or []
var s = 'abcdef';
console.log(s[2]);
console.log(s.substr(2, 1));
console.log(s[2]);
console.log(s.substr(2, 1));
main.prototype.testLoops = function () {
    var x = 2;
    for (var i = 0; i <= 10; i += 2) {
        console.log(i);
    }
    for (i = 0; i >= -10; i -= 2) {
        console.log(i);
    }
    for (var i3 = 0, _stepi3 = x; (10 - i3) * _stepi3 > 0; i3 += _stepi3) {
        console.log(i3);
    }
    for (i3 = 0, _stepi3 = x; (10 - i3) * _stepi3 > 0; i3 += _stepi3) {
        console.log(i3);
    }
    for (var j = 0; j < 3; j++) {
        console.log(j);
    }
    for (j = 0; j < 3; j++) {
        console.log(j);
    }
    for (var _i = 0; _i < 3; _i++) {
        console.log(x);
    }
    for (var k in x) {
        console.log(Object.keys(x));
    }
};

// E02: unknown method: num_tournaments()
// W03: isA('Float',n) is true for all numbers
